src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs

```diff
@@ -44,32 +46,48 @@ public override void Terminate(SubscribeContext context)
 
                 var address = publisherAddress;
 
-                ThreadPool.QueueUserWorkItem(state =>
-                    SendSubscribeMessageWithRetries(address, subscriptionMessage, eventType.AssemblyQualifiedName, context));
+                subscribeTasks.Add(SendSubscribeMessageWithRetries(address, subscriptionMessage, eventType.AssemblyQualifiedName, context));
             }
+
+            return Task.WhenAll(subscribeTasks.ToArray());
         }
 
-        void SendSubscribeMessageWithRetries(string destination, OutgoingMessage subscriptionMessage, string messageType, ContextBag context, int retriesCount = 0)
+        async Task SendSubscribeMessageWithRetries(string destination, OutgoingMessage subscriptionMessage, string messageType, ContextBag context, int retriesCount = 0)
         {
+            var state = context.GetOrCreate<State>();
             try
             {
                 var dispatchOptions = new DispatchOptions(new DirectToTargetDestination(destination), context);
-                dispatcher.Dispatch(new [] { new TransportOperation(subscriptionMessage, dispatchOptions)}).GetAwaiter().GetResult();
+                await dispatcher.Dispatch(new [] { new TransportOperation(subscriptionMessage, dispatchOptions)}).ConfigureAwait(false);
             }
             catch (QueueNotFoundException ex)
             {
-                if (retriesCount < 10)
+                if (retriesCount < state.MaxRetries)
                 {
-                    Thread.Sleep(TimeSpan.FromSeconds(2));
-                    SendSubscribeMessageWithRetries(destination, subscriptionMessage, messageType, context, ++retriesCount);
+                    await Task.Delay(state.RetryDelay).ConfigureAwait(false);
+                    await SendSubscribeMessageWithRetries(destination, subscriptionMessage, messageType, context, ++retriesCount).ConfigureAwait(false);
                 }
                 else
                 {
-                    Logger.ErrorFormat("Failed to subscribe to {0} at publisher queue {1}, reason {2}", messageType, destination, ex.Message);
+                    string message = $"Failed to subscribe to {messageType} at publisher queue {destination}, reason {ex.Message}";
+                    Logger.Error(message, ex);
+                    throw new QueueNotFoundException(destination, message, ex);
                 }
             }
         }
 
+        public class State
```

Comment:
I know we've sort of standardized on `State` but this looks more like `Settings` to me?